### PR TITLE
[SPARK-11797][SQL] collect, first, and take should use encoders for serialization

### PR DESCRIPTION
**Pull request by rxin**
They were previously using Spark's default serializer for serialization.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -529,7 +530,12 @@ class Dataset[T] private[sql](
    * For Java API, use [[collectAsList]].
    * @since 1.6.0
    */
-  def collect(): Array[T] = rdd.collect()
+  def collect(): Array[T] = {
+    val tEnc = resolvedTEncoder
+    val input = queryExecution.analyzed.output
+    val bound = tEnc.bind(input)
+    queryExecution.toRdd.map(_.copy()).collect().map(bound.fromRow)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: hmm, it looks to me that the only difference between this and `Dataset.rdd` is you add a `copy` here, should we remove `Dataset.rdd`?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: the main difference is that we are collecting the rows and then running the encoders to do conversions here.

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
retest this please.

**Comment by rxin** [comment]
I'm going to optimistically merge this since most of the commits passed tests in the other pull request that includes this one.

**Pull request closed without merging**